File: AlgorithmLcr/Try1-Completely unknown single port fitting/cppversion/src/library.cpp
#include "library.hpp"

#include <algorithm>
#include <functional>
#include <mutex>
#include <set>
#include <tuple>

namespace rlc {

namespace {

const TreePtr& leafR() {
    static TreePtr t = Tree::makeLeaf('R');
    return t;
}
const TreePtr& leafL() {
    static TreePtr t = Tree::makeLeaf('L');
    return t;
}
const TreePtr& leafC() {
    static TreePtr t = Tree::makeLeaf('C');
    return t;
}
const std::vector<TreePtr>& theLeaves() {
    static std::vector<TreePtr> v{leafR(), leafL(), leafC()};
    return v;
}

// Leaf admissibility among the direct children of a node (R2' + R4):
//  - PAR: at most one leaf per kind EXCEPT L (parallel (L + Rd) devices are
//    second-order tanks, not mergeable);
//  - SER: all leaf kinds unique, and R must not coexist with L (R4: the
//    series R folds into the inductor's DC resistance).
bool childrenOk(NK kind, const std::vector<TreePtr>& children) {
    int mask = 0;  // bits: 0=R, 1=L, 2=C
    for (const auto& c : children) {
        if (!c->isLeaf) continue;
        int bit = 1 << (c->elem == 'R' ? 0 : c->elem == 'L' ? 1 : 2);
        if (kind == NK::Par && c->elem == 'L') continue;  // multi-L allowed
        if (mask & bit) return false;
        mask |= bit;
    }
    if (kind == NK::Ser) {
        return !((mask & 1) && (mask & 2));  // R4: no R together with L
    }
    return true;
}

struct GenCache {
    std::mutex mu;
    // key: (n, parent_kind, max_idepth) -> subtree options
    std::map<std::tuple<int, int, int>, std::vector<TreePtr>> subtreeOptions;
    // key: (n, max_idepth) -> trees of size n
    std::map<std::pair<int, int>, std::vector<TreePtr>> treesOfSize;
    // key: (max_n, max_idepth) -> cumulative library
    std::map<std::pair<int, int>, std::vector<TreePtr>> libraries;
};

GenCache& cache() {
    static GenCache c;
    return c;
}

std::vector<TreePtr> rootedTrees(int n, NK kind, int max_idepth);

// All canonical subtrees with `size` leaves attachable under a `parent_kind`
// node: leaves at size 1, otherwise opposite-kind roots within the depth
// budget.
std::vector<TreePtr> subtreeOptions(int size, NK parentKind, int max_idepth) {
    if (size == 1) return theLeaves();
    if (max_idepth < 1) return {};
    return rootedTrees(size, opposite(parentKind), max_idepth);
}

std::vector<TreePtr> rootedTrees(int n, NK kind, int max_idepth) {
    // options[size] = subtrees of that size attachable under this node
    std::vector<std::vector<TreePtr>> options;  // index size-1
    std::vector<bool> hasOption;
    for (int size = 1; size < n; ++size) {
        auto opts = subtreeOptions(size, kind, max_idepth - 1);
        hasOption.push_back(!opts.empty());
        options.push_back(std::move(opts));
    }

    std::vector<TreePtr> results;
    std::vector<TreePtr> chosen;

    std::function<void(int)> dfs = [&](int remaining) {
        if (remaining == 0) {
            if (chosen.size() >= 2 && childrenOk(kind, chosen)) {
                results.push_back(Tree::makeNode(kind, chosen));
            }
            return;
        }
        // candidate next child: canonical order >= previous (multiset order)
        std::string start;
        if (!chosen.empty()) start = canonical(chosen.back());
        for (int size = 1; size < n; ++size) {
            if (!hasOption[size - 1]) continue;
            if (size > remaining) continue;
            for (const auto& opt : options[size - 1]) {
                if (!start.empty() && canonical(opt) < start) continue;
                chosen.push_back(opt);
                dfs(remaining - size);
                chosen.pop_back();
            }
        }
    };
    dfs(n);

    // the multiset DFS is duplicate-free by construction; dedup defensively
    std::set<std::string> seen;
    std::vector<TreePtr> uniq;
    for (auto& t : results) {
        auto cs = canonical(t);
        if (seen.insert(cs).second) uniq.push_back(t);
    }
    return uniq;
}

std::vector<TreePtr> treesOfSizeUncached(int n, int max_idepth) {
    if (n == 1) return theLeaves();
    std::vector<TreePtr> out;
    for (NK kind : {NK::Ser, NK::Par}) {
        auto v = rootedTrees(n, kind, max_idepth);
        out.insert(out.end(), v.begin(), v.end());
    }
    std::stable_sort(out.begin(), out.end(),
                     [](const TreePtr& a, const TreePtr& b) {
                         return canonical(a) < canonical(b);
                     });
    return out;
}

// caller must hold cache().mu
std::vector<TreePtr> treesOfSizeLocked(int n, int max_idepth) {
    auto& c = cache();
    auto key = std::make_pair(n, max_idepth);
    auto it = c.treesOfSize.find(key);
    if (it != c.treesOfSize.end()) return it->second;
    auto v = treesOfSizeUncached(n, max_idepth);
    return c.treesOfSize.emplace(key, std::move(v)).first->second;
}

}  // namespace

const std::vector<TreePtr>& TopologyLibrary::get(int max_n, int max_idepth) {
    auto& c = cache();
    std::lock_guard<std::mutex> lock(c.mu);
    auto key = std::make_pair(max_n, max_idepth);
    auto it = c.libraries.find(key);
    if (it != c.libraries.end()) return it->second;
    std::vector<TreePtr> out;
    for (int n = 1; n <= max_n; ++n) {
        auto v = treesOfSizeLocked(n, max_idepth);
        out.insert(out.end(), v.begin(), v.end());
    }
    return c.libraries.emplace(key, std::move(out)).first->second;
}

int TopologyLibrary::countOfSize(int n, int max_idepth) {
    auto& c = cache();
    std::lock_guard<std::mutex> lock(c.mu);
    return (int)treesOfSizeLocked(n, max_idepth).size();
}

std::vector<TreePtr> TopologyLibrary::ofSize(int n, int max_idepth) {
    if (n < 1) throw std::invalid_argument("device count must be >= 1");
    auto& c = cache();
    std::lock_guard<std::mutex> lock(c.mu);
    return treesOfSizeLocked(n, max_idepth);
}

void TopologyLibrary::clearCache() {
    auto& c = cache();
    std::lock_guard<std::mutex> lock(c.mu);
    c.subtreeOptions.clear();
    c.treesOfSize.clear();
    c.libraries.clear();
}

void warmLibraryCache(int max_n, int max_idepth) {
    TopologyLibrary::get(max_n, max_idepth);
}

}  // namespace rlc

File: AlgorithmLcr/Try1-Completely unknown single port fitting/rlc_id/library.py
"""Canonical series-parallel topology enumerator (DESIGN.md section 4.2).

Recursive generation following appendix B.1: split the leaf budget into
integer partitions, recurse with alternating node kinds (R1), enforce the
per-node leaf admissibility rules (R2' + R4, see ``_children_ok``) and store
children in canonical order (R3), so deduplication reduces to a
canonical-string lookup.

Depth limit (deviation note): the generator defaults to ``max_idepth=2``,
i.e. non-root internal nodes have leaf-only children.  Pass
``max_idepth>=3`` to ``library``/``get_library`` to enumerate deeper trees.

Measured counts at the default depth (locked by tests):
  n = 1..6  ->  3, 6, 22, 45, 87, 162   (depth 3 at n=4: 99)
(v2 model: parallel (L + Rd) tanks are distinct devices, so PAR nodes may
hold several L leaves; series R + L collapses to one L device via R4.)
Compare v1 (ideal L): 3, 6, 20, 36, 54, 78 -- the delta are the new
parallel-multi-L topologies, minus the absorbed series R+L forms.
"""

from __future__ import annotations

from functools import lru_cache

from .circuits import SER, PAR, Leaf, Node, Tree, canonical, make_node, n_leaves, opposite

DEFAULT_MAX_IDEPTH = 2

_LEAVES = (Leaf("R"), Leaf("L"), Leaf("C"))


def _children_ok(kind: str, children: list[Tree]) -> bool:
    """Leaf admissibility among the direct children of a ``kind`` node.

    R2': at most one leaf of each kind -- except that a PAR node may hold
         several L leaves (two parallel (L + Rd) devices are a second-order
         tank, not mergeable; series L leaves DO merge, so SER keeps at
         most one L leaf).
    R4:  a SER node never holds both an R leaf and an L leaf (the series R
         folds into the inductor's DC resistance -- one device).
    """
    leaf_kinds = [c.kind for c in children if isinstance(c, Leaf)]
    if kind == PAR:
        non_l = [k for k in leaf_kinds if k != "L"]
        return len(non_l) == len(set(non_l))
    # SER: all leaf kinds unique ...
    if len(leaf_kinds) != len(set(leaf_kinds)):
        return False
    # ... and R must not coexist with L (R4 series absorption)
    return not ("R" in leaf_kinds and "L" in leaf_kinds)


@lru_cache(maxsize=None)
def _subtree_options(size: int, parent_kind: str, max_idepth: int) -> tuple[Tree, ...]:
    """All canonical subtrees with ``size`` leaves attachable under a node of
    ``parent_kind``; may include leaves (size 1) or internal nodes of the
    opposite kind within the remaining internal-depth budget."""
    if size == 1:
        return _LEAVES
    if max_idepth < 1:
        return ()
    kind = opposite(parent_kind)
    return tuple(_rooted_trees(size, kind, max_idepth))


def _rooted_trees(n: int, kind: str, max_idepth: int) -> list[Tree]:
    """All canonical trees of size n whose root node has the given kind."""
    # options[size] = subtrees of that size attachable under this node
    options: dict[int, tuple[Tree, ...]] = {}
    for size in range(1, n):
        opts = _subtree_options(size, kind, max_idepth - 1)
        if opts:
            options[size] = opts

    results: list[Tree] = []

    def dfs(remaining: int, chosen: list[Tree]) -> None:
        if remaining == 0:
            if len(chosen) >= 2 and _children_ok(kind, chosen):
                results.append(make_node(kind, chosen))
            return
        # candidate next child: canonical order >= previous (multiset order)
        start = canonical(chosen[-1]) if chosen else ""
        for size, opts in options.items():
            if size > remaining:
                continue
            for opt in opts:
                if start and canonical(opt) < start:
                    continue
                dfs(remaining - size, chosen + [opt])

    dfs(n, [])
    # The multiset DFS is duplicate-free by construction; dedup defensively.
    seen: set[str] = set()
    uniq: list[Tree] = []
    for t in results:
        cs = canonical(t)
        if cs not in seen:
            seen.add(cs)
            uniq.append(t)
    return uniq


@lru_cache(maxsize=None)
def _trees_of_size(n: int, max_idepth: int) -> tuple[Tree, ...]:
    """All canonical topologies with exactly n devices."""
    if n == 1:
        return _LEAVES
    out: list[Tree] = []
    for kind in (SER, PAR):
        out.extend(_rooted_trees(n, kind, max_idepth))
    return tuple(sorted(out, key=canonical))


def trees_of_size(n: int, max_idepth: int = DEFAULT_MAX_IDEPTH) -> tuple[Tree, ...]:
    """Public single-layer view: all canonical topologies with exactly ``n``
    devices (the layer used by the optional exact-count constraint)."""
    if n < 1:
        raise ValueError(f"device count must be a positive integer, got {n}")
    return _trees_of_size(n, max_idepth)


@lru_cache(maxsize=None)
def get_library(max_n: int = 4, max_idepth: int = DEFAULT_MAX_IDEPTH) -> tuple[Tree, ...]:
    """All canonical topologies with 1..max_n devices (cached)."""
    out: list[Tree] = []
    for n in range(1, max_n + 1):
        out.extend(_trees_of_size(n, max_idepth))
    return tuple(out)


def counts(max_n: int = 6, max_idepth: int = DEFAULT_MAX_IDEPTH) -> dict[int, int]:
    """Number of canonical topologies for each device count 1..max_n."""
    return {n: len(_trees_of_size(n, max_idepth)) for n in range(1, max_n + 1)}


def stats(max_n: int = 6, max_idepth: int = DEFAULT_MAX_IDEPTH) -> str:
    """Print the canonical-topology count table (n = 1..max_n)."""
    cs = counts(max_n, max_idepth)
    lines = [
        f"canonical series-parallel topologies (max internal depth = {max_idepth})",
        "+-----+-----------------+",
        "|   n | topologies      |",
        "+-----+-----------------+",
    ]
    for n, c in cs.items():
        lines.append(f"| {n:3d} | {c:15d} |")
    lines.append("+-----+-----------------+")
    text = "\n".join(lines)
    print(text)
    return text


__all__ = ["get_library", "trees_of_size", "counts", "stats",
           "DEFAULT_MAX_IDEPTH", "n_leaves"]
